### ISSUE TITLE
Using with_fileglob and "delegate_to: localhost" causes ansible to make ssh connection to localhost

### ISSUE THREAD

**Issue by mwoodson** [ansible/ansible#18549]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

with_fileglob

##### ANSIBLE VERSION

```
ansible 2.2.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### CONFIGURATION

N/A

##### OS / ENVIRONMENT

Linux - RHEL 7

##### SUMMARY

This is appears to be fairly specific, but there is an issue with running with_fileglob and "delegate_to: localhost".  Normally, when delegate_to: localhost is called, it will use the "ansible_connection: local".  But, with the combination mentioned, it is trying to do an SSH connection to localhost.

The weird thing, if I change the code and remove "delegate_to: localhost", or change with_fileglob to with_items, I no longer see the ssh execution.

Please see the steps to reproduce.  This might easier to explain and show in example.

This isn't normally a big deal, but when trying to run ansible from within a docker container, this tries to do an SSH to the localhost running docker, not the actual docker container itself.

##### STEPS TO REPRODUCE

```
$ touch foobar
$ cat test.yml 
- hosts: localhost
  gather_facts: no

  pre_tasks:
  - stat:
      path: "{{ item }}"
    with_fileglob:
    - foo*
    delegate_to: localhost
```

when run, notice the SSH that it tries to use:

```
$ ansible-playbook -vvvvv test.yml
Using /etc/ansible/ansible.cfg as config file
Loading callback plugin default of type stdout, v2.0 from /usr/lib/python2.7/site-packages/ansible/plugins/callback/__init__.pyc

PLAYBOOK: test.yml *************************************************************
1 plays in test.yml

PLAY [localhost] ***************************************************************
looking for "" at "/files/"
looking for "" at "/"

TASK [stat] ********************************************************************
task path: /home/mwoodson/tmp/playbooks/test.yml:5
looking for "" at "/home/mwoodson/tmp/playbooks/files/"
looking for "" at "/home/mwoodson/tmp/playbooks/"
Using module file /usr/lib/python2.7/site-packages/ansible/modules/core/files/stat.py
<localhost> ESTABLISH SSH CONNECTION FOR USER: None
<localhost> SSH: ansible.cfg set ssh_args: (-o)(ControlMaster=auto)(-o)(ControlPersist=600s)
<localhost> SSH: ANSIBLE_HOST_KEY_CHECKING/host_key_checking disabled: (-o)(StrictHostKeyChecking=no)
<localhost> SSH: ansible_password/ansible_ssh_pass not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<localhost> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<localhost> SSH: PlayContext set ssh_common_args: ()
<localhost> SSH: PlayContext set ssh_extra_args: ()
<localhost> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/mwoodson/.ansible/cp/ansible-ssh-%h-%p-%r)
<localhost> SSH: EXEC ssh -vvv -o ControlMaster=auto -o ControlPersist=600s -o StrictHostKeyChecking=no -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/mwoodson/.ansible/cp/ansible-ssh-%h-%p-%r localhost '/bin/sh -c '"'"'/usr/bin/python && sleep 0'"'"''
ok: [localhost] => (item=/home/mwoodson/tmp/playbooks/foobar) => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "checksum_algorithm": "sha1", 
            "follow": false, 
            "get_checksum": true, 
            "get_md5": true, 
            "mime": false, 
            "path": "/home/mwoodson/tmp/playbooks/foobar"
        }, 
        "module_name": "stat"
    }, 
    "item": "/home/mwoodson/tmp/playbooks/foobar", 
    "stat": {
        "atime": 1479491278.491219, 
        "checksum": "da39a3ee5e6b4b0d3255bfef95601890afd80709", 
...
}

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0   

```

##### EXPECTED RESULTS

I would expect it to run on localhost, and not cause an SSH connection.

##### ACTUAL RESULTS

```
<localhost> ESTABLISH SSH CONNECTION FOR USER: None
<localhost> SSH: ansible.cfg set ssh_args: (-o)(ControlMaster=auto)(-o)(ControlPersist=600s)
<localhost> SSH: ANSIBLE_HOST_KEY_CHECKING/host_key_checking disabled: (-o)(StrictHostKeyChecking=no)

```

**Comment by alikins** [comment]
I can reproduce this from devel/ branch:
(using playbook from https://github.com/alikins/ansible-bug-repro/tree/master/with_fileglob_ssh_18549)

```
[fedora-23:ansible (devel % u=)]$ cat repro/with_fileglob_ssh_18549/hosts 
localhost ansible_connection=local
local_alias ansible_connection=local
[fedora-23:ansible (devel % u=)]$ cat repro/with_fileglob_ssh_18549/test.yml 
- hosts: localhost
  gather_facts: no

  pre_tasks:
  - stat:
      path: "{{ item }}"
    with_fileglob:
      - "foo*"
    delegate_to: localhost
    register: blip

  #  tasks:
  #  - debug: var=blip
[fedora-23:ansible (devel % u=)]$ ansible --version
ansible 2.3.0 (devel 6dece90a57) last updated 2016/11/21 10:41:54 (GMT -400)
  lib/ansible/modules/core: (detached HEAD 5c986306be) last updated 2016/11/21 09:38:36 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 9511de1e3d) last updated 2016/11/18 14:14:43 (GMT -400)
  config file = /home/adrian/.ansible.cfg
  configured module search path = Default w/o overrides
[fedora-23:ansible (devel % u=)]$ ansible-playbook -vvvvv -i repro/with_fileglob_ssh_18549/hosts repro/with_fileglob_ssh_18549/test.yml 
Using /home/adrian/.ansible.cfg as config file
Loading callback plugin default of type stdout, v2.0 from /home/adrian/src/ansible/lib/ansible/plugins/callback/__init__.pyc

PLAYBOOK: test.yml ************************************************************************************************************************************************************************************************
1 plays in repro/with_fileglob_ssh_18549/test.yml

PLAY [localhost] **************************************************************************************************************************************************************************************************
looking for "" at "/files/"
looking for "" at "/"

TASK [stat] *******************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible/repro/with_fileglob_ssh_18549/test.yml:5
looking for "" at "/home/adrian/src/ansible-bug-repro/with_fileglob_ssh_18549/files/"
looking for "" at "/home/adrian/src/ansible-bug-repro/with_fileglob_ssh_18549/"
Using module file /home/adrian/src/ansible/lib/ansible/modules/core/files/stat.py
<localhost> ESTABLISH SSH CONNECTION FOR USER: None
<localhost> SSH: ansible.cfg set ssh_args: (-C)(-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<localhost> SSH: ansible_password/ansible_ssh_pass not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<localhost> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<localhost> SSH: PlayContext set ssh_common_args: ()
<localhost> SSH: PlayContext set ssh_extra_args: ()
<localhost> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r)
<localhost> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r localhost '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo ~/.ansible/tmp/ansible-tmp-1479744843.69-232995974026075 `" && echo ansible-tmp-1479744843.69-232995974026075="` echo ~/.ansible/tmp/ansible-tmp-1479744843.69-232995974026075 `" ) && sleep 0'"'"''
<localhost> PUT /tmp/tmp2miGiK TO /home/adrian/.ansible/tmp/ansible-tmp-1479744843.69-232995974026075/stat.py
<localhost> SSH: ansible.cfg set ssh_args: (-C)(-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<localhost> SSH: ansible_password/ansible_ssh_pass not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<localhost> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<localhost> SSH: PlayContext set ssh_common_args: ()
<localhost> SSH: PlayContext set sftp_extra_args: ()
<localhost> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r)
<localhost> SSH: EXEC sftp -b - -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r '[localhost]'
<localhost> ESTABLISH SSH CONNECTION FOR USER: None
<localhost> SSH: ansible.cfg set ssh_args: (-C)(-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<localhost> SSH: ansible_password/ansible_ssh_pass not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<localhost> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<localhost> SSH: PlayContext set ssh_common_args: ()
<localhost> SSH: PlayContext set ssh_extra_args: ()
<localhost> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r)
<localhost> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r localhost '/bin/sh -c '"'"'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1479744843.69-232995974026075/ /home/adrian/.ansible/tmp/ansible-tmp-1479744843.69-232995974026075/stat.py && sleep 0'"'"''
<localhost> ESTABLISH SSH CONNECTION FOR USER: None
<localhost> SSH: ansible.cfg set ssh_args: (-C)(-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<localhost> SSH: ansible_password/ansible_ssh_pass not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<localhost> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<localhost> SSH: PlayContext set ssh_common_args: ()
<localhost> SSH: PlayContext set ssh_extra_args: ()
<localhost> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r)
<localhost> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r -tt localhost '/bin/sh -c '"'"'/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1479744843.69-232995974026075/stat.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1479744843.69-232995974026075/" > /dev/null 2>&1 && sleep 0'"'"''
ok: [localhost] => (item=/home/adrian/src/ansible-bug-repro/with_fileglob_ssh_18549/foobar) => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "checksum_algorithm": "sha1", 
            "follow": false, 
            "get_attributes": true, 
            "get_checksum": true, 
            "get_md5": true, 
            "get_mime": true, 
            "path": "/home/adrian/src/ansible-bug-repro/with_fileglob_ssh_18549/foobar"
        }, 
        "module_name": "stat"
    }, 
    "item": "/home/adrian/src/ansible-bug-repro/with_fileglob_ssh_18549/foobar", 
    "stat": {
        "atime": 1479743329.7910044, 
        "attr_flags": "e", 
        "attributes": [
            "extents"
        ], 
        "block_size": 4096, 
        "blocks": 0, 
        "charset": "binary", 
        "checksum": "da39a3ee5e6b4b0d3255bfef95601890afd80709", 
        "ctime": 1479743240.2088218, 
        "dev": 64770, 
        "device_type": 0, 
        "executable": false, 
        "exists": true, 
        "gid": 1000, 
        "gr_name": "adrian", 
        "inode": 4205065, 
        "isblk": false, 
        "ischr": false, 
        "isdir": false, 
        "isfifo": false, 
        "isgid": false, 
        "islnk": false, 
        "isreg": true, 
        "issock": false, 
        "isuid": false, 
        "md5": "d41d8cd98f00b204e9800998ecf8427e", 
        "mimetype": "inode/x-empty", 
        "mode": "0664", 
        "mtime": 1479743240.2088218, 
        "nlink": 1, 
        "path": "/home/adrian/src/ansible-bug-repro/with_fileglob_ssh_18549/foobar", 
        "pw_name": "adrian", 
        "readable": true, 
        "rgrp": true, 
        "roth": true, 
        "rusr": true, 
        "size": 0, 
        "uid": 1000, 
        "version": "18446744072465469325", 
        "wgrp": true, 
        "woth": false, 
        "writeable": true, 
        "wusr": true, 
        "xgrp": false, 
        "xoth": false, 
        "xusr": false
    }
}

PLAY RECAP ********************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0   
```

**Comment by alikins** [comment]
repro using c&p or play from report:

```
[fedora-23:ansible (devel % u=)]$ cat repro/with_fileglob_ssh_18549/hosts
localhost ansible_connection=local
other ansible_connection=local
#local_alias ansible_connection=local
[fedora-23:ansible (devel % u=)]$ cat repro/with_fileglob_ssh_18549/orig.yml 
- hosts: localhost
  gather_facts: no

  pre_tasks:
  - stat:
      path: "{{ item }}"
    with_fileglob:
    - foo*
    delegate_to: localhost
[fedora-23:ansible (devel % u=)]$ ansible-playbook -vvvvv -i repro/with_fileglob_ssh_18549/hosts repro/with_fileglob_ssh_18549/orig.yml 
Using /home/adrian/.ansible.cfg as config file
Loading callback plugin default of type stdout, v2.0 from /home/adrian/src/ansible/lib/ansible/plugins/callback/__init__.pyc

PLAYBOOK: orig.yml ************************************************************************************************************************************************************************************************
1 plays in repro/with_fileglob_ssh_18549/orig.yml

PLAY [localhost] **************************************************************************************************************************************************************************************************
looking for "" at "/files/"
looking for "" at "/"

TASK [stat] *******************************************************************************************************************************************************************************************************
task path: /home/adrian/src/ansible/repro/with_fileglob_ssh_18549/orig.yml:5
looking for "" at "/home/adrian/src/ansible-bug-repro/with_fileglob_ssh_18549/files/"
looking for "" at "/home/adrian/src/ansible-bug-repro/with_fileglob_ssh_18549/"
Using module file /home/adrian/src/ansible/lib/ansible/modules/core/files/stat.py
<localhost> ESTABLISH SSH CONNECTION FOR USER: None
<localhost> SSH: ansible.cfg set ssh_args: (-C)(-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<localhost> SSH: ansible_password/ansible_ssh_pass not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<localhost> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<localhost> SSH: PlayContext set ssh_common_args: ()
<localhost> SSH: PlayContext set ssh_extra_args: ()
<localhost> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r)
<localhost> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r localhost '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo ~/.ansible/tmp/ansible-tmp-1479746224.94-228087769325633 `" && echo ansible-tmp-1479746224.94-228087769325633="` echo ~/.ansible/tmp/ansible-tmp-1479746224.94-228087769325633 `" ) && sleep 0'"'"''
<localhost> PUT /tmp/tmpPCaYdd TO /home/adrian/.ansible/tmp/ansible-tmp-1479746224.94-228087769325633/stat.py
<localhost> SSH: ansible.cfg set ssh_args: (-C)(-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<localhost> SSH: ansible_password/ansible_ssh_pass not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<localhost> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<localhost> SSH: PlayContext set ssh_common_args: ()
<localhost> SSH: PlayContext set sftp_extra_args: ()
<localhost> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r)
<localhost> SSH: EXEC sftp -b - -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r '[localhost]'
<localhost> ESTABLISH SSH CONNECTION FOR USER: None
<localhost> SSH: ansible.cfg set ssh_args: (-C)(-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<localhost> SSH: ansible_password/ansible_ssh_pass not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<localhost> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<localhost> SSH: PlayContext set ssh_common_args: ()
<localhost> SSH: PlayContext set ssh_extra_args: ()
<localhost> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r)
<localhost> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r localhost '/bin/sh -c '"'"'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1479746224.94-228087769325633/ /home/adrian/.ansible/tmp/ansible-tmp-1479746224.94-228087769325633/stat.py && sleep 0'"'"''
<localhost> ESTABLISH SSH CONNECTION FOR USER: None
<localhost> SSH: ansible.cfg set ssh_args: (-C)(-o)(ControlMaster=auto)(-o)(ControlPersist=60s)
<localhost> SSH: ansible_password/ansible_ssh_pass not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<localhost> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<localhost> SSH: PlayContext set ssh_common_args: ()
<localhost> SSH: PlayContext set ssh_extra_args: ()
<localhost> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r)
<localhost> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r -tt localhost '/bin/sh -c '"'"'/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1479746224.94-228087769325633/stat.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1479746224.94-228087769325633/" > /dev/null 2>&1 && sleep 0'"'"''
ok: [localhost] => (item=/home/adrian/src/ansible-bug-repro/with_fileglob_ssh_18549/foobar) => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "checksum_algorithm": "sha1", 
            "follow": false, 
            "get_attributes": true, 
            "get_checksum": true, 
            "get_md5": true, 
            "get_mime": true, 
            "path": "/home/adrian/src/ansible-bug-repro/with_fileglob_ssh_18549/foobar"
        }, 
        "module_name": "stat"
    }, 
    "item": "/home/adrian/src/ansible-bug-repro/with_fileglob_ssh_18549/foobar", 
    "stat": {
        "atime": 1479743329.7910044, 
        "attr_flags": "e", 
        "attributes": [
            "extents"
        ], 
        "block_size": 4096, 
        "blocks": 0, 
        "charset": "binary", 
        "checksum": "da39a3ee5e6b4b0d3255bfef95601890afd80709", 
        "ctime": 1479743240.2088218, 
        "dev": 64770, 
        "device_type": 0, 
        "executable": false, 
        "exists": true, 
        "gid": 1000, 
        "gr_name": "adrian", 
        "inode": 4205065, 
        "isblk": false, 
        "ischr": false, 
        "isdir": false, 
        "isfifo": false, 
        "isgid": false, 
        "islnk": false, 
        "isreg": true, 
        "issock": false, 
        "isuid": false, 
        "md5": "d41d8cd98f00b204e9800998ecf8427e", 
        "mimetype": "inode/x-empty", 
        "mode": "0664", 
        "mtime": 1479743240.2088218, 
        "nlink": 1, 
        "path": "/home/adrian/src/ansible-bug-repro/with_fileglob_ssh_18549/foobar", 
        "pw_name": "adrian", 
        "readable": true, 
        "rgrp": true, 
        "roth": true, 
        "rusr": true, 
        "size": 0, 
        "uid": 1000, 
        "version": "18446744072465469325", 
        "wgrp": true, 
        "woth": false, 
        "writeable": true, 
        "wusr": true, 
        "xgrp": false, 
        "xoth": false, 
        "xusr": false
    }
}

PLAY RECAP ********************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0   
```

**Comment by siniG** [comment]
Hi,

We ran into the same problem when running with Centos 6.5 and 6.8.
We solved it by using with_items and using a lookup query for the files under the directory we needed

**Comment by sivel** [comment]
I've been doing a little bit of research on this, and here is what I have so far.

When running `VariableManager.get_vars` we populate `ansible_delegated_vars` via `VariableManager._get_delegated_vars`.  When this runs in one instance, `ansible_search_path` is undefined.  Due to this, when using relative paths in `with_fileglob`, the `fileglob` lookup returns no results.

Because `fileglob` returns no results due to `ansible_search_path` being undefined, `VariableManager._get_delegated_vars` doesn't set `ansible_delegated_vars`, and due to that, we have no specified connection vars, so we default to `ssh` connection.

I haven't read through all the code yet to understand why, but we are effectively calling the lookup multiple times, both with and without proper context, and the one with improper context causes things to go bad.

**Comment by sivel** [comment]
~~To add to this, it only happens when the first task of the play uses `with_fileglob`.  Any other task as the first task, without `with_fileglob`, and the play will execute as expected.~~

My mistake, the first task comment, is false, I had tried so many things, and had converted my paths to full.

**Comment by agaffney** [comment]
Nice find. This sounds vaguely reminiscent of another issue I was helping a user with earlier in the 2.x line where all `lookup` calls in a vars file were being evaluated at least once, instead of only when the var that references them was itself referenced, as was the case in the previous version.

Edit: I think this particular comment is what I am thinking of

https://github.com/ansible/ansible/issues/17001#issuecomment-238341044

**Comment by sivel** [comment]
It looks like why this is being done, is to template the `delegate_to` directive, when a loop is used.  Unfortunately, since we use `find_file_in_search_path` in `with_fileglob` along with the `ansible_search_path` value, when we run this early, without full context, we get nothing from the `fileglob` lookup, and as such, never update `ansible_delegated_vars`.

The following seems to resolve the issue:

```diff
diff --git a/lib/ansible/vars/__init__.py b/lib/ansible/vars/__init__.py
index dfa3131..9cb1072 100644
--- a/lib/ansible/vars/__init__.py
+++ b/lib/ansible/vars/__init__.py
@@ -465,6 +465,11 @@ class VariableManager:
         else:
             items = [None]
 
+        if not items:
+            # We may not have gotten any items back from the lookup, per above
+            # we set up a dummy array so that bad things don't happen
+            items.append(None)
+
         delegated_host_vars = dict()
         for item in items:
             # update the variables with the item value for templating, in case we need it
```

**Comment by sivel** [comment]
I suppose there are a few ways to go here.  Make sure we pass full context for templating `delegate_to`, or the way I follow above.

I guess the problem with the above, is if you are storing `delegate_to` information in files, and use a `with_fileglob` for actually templating `delegate_to`.  In which case, fixing the context thing would be best, but I don't know where that would need to be done as of yet.

**Comment by sivel** [comment]
Sorry for all the comments, just trying to document information needed for resolving this.

See https://github.com/ansible/ansible/pull/16630 / https://github.com/ansible/ansible/commit/3c39bb56 for when I believe this bug was introduced

**Comment by sivel** [comment]
I haven't looked in depth, but we should see if https://github.com/ansible/ansible/pull/19297 helps with this issue.  I'm not sure it does, but I saw components that deal with pathing.

**Comment by sivel** [comment]
This issue was resolved in 9339b0f4a7b9c2cd718e1f4e3476bbe04181f1b2 and subsequently included in the 2.4.0.0-1 release.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project
